include/realizations/catchment/Formulation_Manager.hpp

```diff
@@ -285,16 +293,29 @@ namespace realization {
              * 
              * @return std::string of the output root directory
              */
-            std::string get_output_root() const noexcept {
+            std::string get_output_root() const {
                 const auto output_root = this->tree.get_optional<std::string>("output_root");
                 if (output_root != boost::none && *output_root != "") {
                     // Check if the path ends with a trailing slash,
                     // otherwise add it.
-                    return output_root->back() == '/'
+                    std::string str = output_root->back() == '/'
                            ? *output_root
                            : *output_root + "/";
-                }
 
+                    const char* dir = str.c_str();
+
+                    //use C++ system function to check if there is a dir match that defined in realization
+                    struct stat sb;
+                    if (stat(dir, &sb) == 0 && S_ISDIR(sb.st_mode))
+                        return dir;
+                    else {
+                        //throw std::runtime_error("output_root directory does not exist, please create one matching that in realization");
+                        mkdir(dir, 0755);      
```

Comment:
Two things for `mkdir` that I think we need to ensure.

1. What happens when `mkdir` fails? (i.e. the directory where the requested path needs to be created isn't writable). I think we should check the [return code](https://pubs.opengroup.org/onlinepubs/007904875/functions/mkdir.html) and propagate the error code as a `std::runtime_error`. Additionally, you could skip the `stat` check in favor of checking `mkdir`'s return code for `EEXISTS`, but that's not super necessary.

2. What happens when the requested path requires multiple levels of directories? i.e. given a path `/path/to/ngen` with an `output_root` set to `/path/to/ngen/a/b`, does `mkdir` successfully create `a/` and `a/b/`? I suspect this will fail, so we might want to create a test case to check that.

Reply:
(1) We probably should put in a guard there when mkdir fails. Throwing a `std::runtime_error` would be a good option. I am reasonably happy with the stat check the way it is, unless you feel strongly about it.

(2) My initial research suggests `mkdir `does have the capability to generate multi-level directories. However, for our purpose, do we really have a need for that?

Reply:
For (1), as long as we throw an error, that's fine.

For (2), where are you seeing that `mkdir` can create multi-level directories? 

This example results in an error:
```c
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <errno.h>
#include <unistd.h>

void test_path(const char* path) {
    errno = 0;
    int result_code = mkdir(path, 0775);
    if (result_code == -1)
        printf("[%s] error: %s\n", path, strerror(errno));

    unlink(path);
}

int main(int argc, const char* argv[]) {
    // This passes (unless current directory is not writable)
    test_path("example_path");

    // This fails unless 'example_nested_path/' is created
    test_path("example_nested_path/subdir");
    
    // This fails unless '/opt/ngen/outputs/' is created
    test_path("/opt/ngen/outputs/22July2024_1");
    return 0;
}
```

For your second point in (2), the purpose is for consumers. It's a reasonable assumption that if `output_root` is set to a directory, and it can create a top-level directory, then it should be able to create multi-level directories. However, worst-case, as long as we're throwing the errors, then the consumer will be informed.